### PR TITLE
Reorganize playwright utils and make canvas utils usable in the site editor

### PR DESCRIPTION
**Pull request by talldan**
## What?
related - https://github.com/WordPress/gutenberg/issues/38851#issuecomment-1102003820.

Reorganizes the playwright e2e test utils.

Also converts a lot more of them to TypeScript.

## Why?
When migrating some site editor tests to playwright, I realised that it wasn't straightforward. A lot of existing e2e utils couldn't be used because they're unable to select from the site-editor's iframed editor canvas. Selecting from the iframe canvas requires using `frame` or `frameLocator` instead of `page`.

Puppeteer tests seem to have the same issue, and it resulted in a lot of utils being duplicated for the site editor, often these functions are just declared in the test file.

In this PR i'm trying to get ahead of the issue for the Playwright utils by organizing them in a way that works from the start.

## How?
I've divided up the utils into more granular parts:
- `Admin` - Any utils related to WP Admin (e.g. `visitAdminPage`)
- `Editor` - Any utils related to the block editor (e.g. `insertBlock`). Exposes a `canvas` property that can be used to select elements within the site editor's iframe. This has to be initialised within each test, and we pass a `hasIframe` prop when instantiating for the site editor.
- `PageUtils` - This was previously where all the utils (apart from RequestUtils) were defined, but now this is a place for any generic browser interactions (e.g. `pressKeyWithModifier`).

### CODE REVIEW

File: test/e2e/specs/editor/blocks/paragraph.spec.js [hunk 1]
```diff
@@ -1,18 +1,28 @@
...
+	test,
+	expect,
+	Editor,
+} = require( '@wordpress/e2e-test-utils-playwright' );
+
+test.use( {
+	editor: async ( { page }, use ) => {
+		await use( new Editor( { page } ) );
+	},
+} );
```
Comment (talldan) [inline test/e2e/specs/editor/blocks/paragraph.spec.js]: I don't think we actually need to do this in every test. Site editor tests can probably do something like:
`editor.useIframeCanvas()` in a `beforeAll`. 🤔

Comment (talldan) [reply test/e2e/specs/editor/blocks/paragraph.spec.js]: Actually, this doesn't seem possible, there's an error if I try this -  `Error: "context" and "page" fixtures are not supported in beforeAll. Use browser.newContext() instead.`

Comment (kevin940726) [reply test/e2e/specs/editor/blocks/paragraph.spec.js]: Yeah we can't do any of that stuff in `beforeAll` or `afterAll`, that's one of the reasons we need `requestUtils`.

File: packages/e2e-test-utils-playwright/src/editor/index.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	#hasIframe: boolean;
+
+	constructor( { page, hasIframe = false }: EditorConstructorParams ) {
+		this.page = page;
+		this.context = page.context();
+		this.browser = this.context.browser()!;
+		this.#hasIframe = hasIframe;
+	}
+
+	get canvas(): Frame | Page {
```
Comment (kevin940726) [inline packages/e2e-test-utils-playwright/src/editor/index.ts:37]: I wonder if we could also return `PageUtils` or `Editor` here. So that in tests we could do something like:

```js
const { page /* or frame */, pageUtils, editor, ...etc } = editor.canvas;
```

We can then chain all the existing utils with an explicit API. I haven't looked into it enough to know if it's possible though 😅.

**Review by kevin940726 — COMMENTED** [review]
I think overall the reorganization makes sense! Thanks for putting work on this 👍.

I wonder if it would be better to rename `editor` and `admin` to something like `editorUtils` and `adminUtils` though, just to make their names ergonomic.

File: packages/e2e-test-utils-playwright/src/editor/index.ts (hunk 2 above)
Comment (talldan) [reply packages/e2e-test-utils-playwright/src/editor/index.ts:37]: Would you be able to add some more detail on that? Would the goal be to use pageUtils within the editor canvas?

**Comment by talldan** [comment]
> I wonder if it would be better to rename editor and admin to something like editorUtils and adminUtils though, just to make their names ergonomic.

I quite like the shorter names, it feels nice and concise when writing tests, which are often quite verbose.

The thing that I noticed is that `admin` and `editor` are also user roles 😄 

I wonder whether that will be confusing to newcomers.

File: packages/e2e-test-utils-playwright/src/editor/index.ts (hunk 2 above)
Comment (kevin940726) [reply packages/e2e-test-utils-playwright/src/editor/index.ts:37]: Yep, exactly. I think it's still a problem, right? Though it's not really related to this PR though and we can discuss it in another issue :).

**Comment by kevin940726** [comment]
> The thing that I noticed is that admin and editor are also user roles 😄

Ahh, I see! Putting it this way makes more sense, but I'm not sure if it'd be confusing either. I guess it'd still be okay as long as we document it though. Personally, I'm fine with it 👍.

File: packages/e2e-test-utils-playwright/src/editor/index.ts (hunk 2 above)
Comment (talldan) [reply packages/e2e-test-utils-playwright/src/editor/index.ts:37]: It's tricky because we'd have to pass the appropriate `frame` or `page` into `pageUtils`, possibly instantiating it on the fly within the `canvas` getter, and each util would have to be written in a way that takes that into account.

Looking at the existing puppeteer utils, there aren't many functions that would benefit from this. Possibly only `enableObserveFocusLoss` and `disableObserveFocusLoss` because I think they'd have to listen inside the iframe.

It may be easier to do something like `pageUtils.enableObserveFocusLoss( editor.canvas )`.

**Comment by talldan** [comment]
Ok, I've updated the README file to add some details about the changes. I haven't updated the CHANGELOG, as it looks like the package is still private.

All tests passing, so should be good for further feedback or merge.

File: packages/e2e-test-utils-playwright/src/editor/index.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import { selectBlockByClientId } from './select-block-by-client-id';
+import { showBlockToolbar } from './show-block-toolbar';
+import { saveSiteEditorEntities } from './site-editor';
+
+type EditorConstructorProps = {
+	page: Page;
+	hasIframe?: boolean;
+};
+
+export class Editor {
```
Comment (kevin940726) [inline packages/e2e-test-utils-playwright/src/editor/index.ts:24]: Asking each test to manually include `Editor` and initiate it in `test.use` might not be ergonomic for some. I wonder if we could just auto-initiate them in `test.ts` like we do for `pageUtils` and `admin`? The problem is with the `hasIframe` option though.

Apologies if I missed something here, but could we remove the `hasIframe` boolean, and just use the `canvas` getter for such cases? So that we can initiate the `editor` fixture globally and also allow accessing from the iframe canvas in each test.

**Review by kevin940726 — COMMENTED** [review]
This is looking great!

Now that we have `Editor` and `Admin`, I think it might be beneficial if we update the MIGRATION guide too?

**Comment by talldan** [comment]
> Now that we have Editor and Admin, I think it might be beneficial if we update the MIGRATION guide too?

Where does that live? The overview issue seems to link to a blog post, and it didn't have any outdated info.

File: packages/e2e-test-utils-playwright/src/editor/index.ts (hunk 3 above)
Comment (talldan) [reply packages/e2e-test-utils-playwright/src/editor/index.ts:24]: > Asking each test to manually include Editor and initiate it in test.use might not be ergonomic for some

Yeah, that is the downside.

> Apologies if I missed something here, but could we remove the hasIframe boolean, and just use the canvas getter for such cases?

The idea is that we can write `editor` utils that interact with the canvas and internally they can use `this.canvas`. The test configuration determines whether the iframe or the page is used. Removing `hasIframe` makes that difficult to do.

Potentially we could have some separate configuration that determines whether an iframe is used in a test util. For example, another class that just stores configuration? If it didn't rely on `Page` then it could be instantiated in `test.ts` and `hasIframe` could be set in a `beforeAll` only for site editor tests.

**Comment by kevin940726** [comment]
> Where does that live?

In [`test/e2e/MIGRATION.md`](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/MIGRATION.md)

**Comment by talldan** [comment]
> In [test/e2e/MIGRATION.md](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/MIGRATION.md)

Not sure how I missed that. I've updated those docs.

File: packages/e2e-test-utils-playwright/src/editor/index.ts (hunk 3 above)
Comment (kevin940726) [reply packages/e2e-test-utils-playwright/src/editor/index.ts:24]: I see. 🤔

Technically, I think we can still _useCanvas_ in `beforeEach` though? Would that make things easier?

Comment (talldan) [reply packages/e2e-test-utils-playwright/src/editor/index.ts:24]: I think I managed to get it working without needing `Editor` to be setup in every test, the setup is now only needed in site editor tests. If this doesn't work we can try the beforeEach

Comment (kevin940726) [reply packages/e2e-test-utils-playwright/src/editor/index.ts:24]: Cool! Amazingly simple solution 😆.

File: test/e2e/specs/site-editor/template-part.spec.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
+/**
```
Comment (kevin940726) [inline test/e2e/specs/site-editor/template-part.spec.js]: Let's split this into a different PR to better review/manage it?

**Review by kevin940726 — APPROVED** [review]
This looks awesome! I didn't spot anything that's blocking us from merging this so let's do it! 🚢 

Maybe we'd want to remove the "Try" from the PR title though 😆.

File: test/e2e/specs/site-editor/template-part.spec.js (hunk 4 above)
Comment (talldan) [reply test/e2e/specs/site-editor/template-part.spec.js]: Good reminder, I'll remove it.

Thanks for all the review feedback and helping get this to a good place.

**Comment by kevin940726** [comment]
@talldan Since now that Post Editor also uses iframe, would it make sense to default the `hasIframe` option to `true` to eliminate some of the setups?

Also, curious to ask, could a simple check in the `canvas` getter get rid of the `hasIframe` option altogether?

```ts
get canvas(): Frame | Page {
	let frame = this.page.frame( 'editor-canvas' ) || this.page;

	if ( ! frame ) {
		throw new Error(
			'EditorUtils: unable to find editor canvas iframe or page'
		);
	}

	return frame;
}
```

**Comment by talldan** [comment]
@kevin940726 Sounds like a good idea. 

> Also, curious to ask, could a simple check in the canvas getter get rid of the hasIframe option altogether?

If it works, that'd be great. I think the 'editor-canvas' class still existed as a `div` element previously, so as long as `this.page.frame( 'editor-canvas' )` doesn't do something weird when that's the case that would seem like a good option.
